libass/ass_fontselect.c

```diff
@@ -256,6 +260,404 @@ static ASS_FontInfo *allocate_font_info(ASS_FontSelector *selector)
+ *
+ * \param addr address to read from
+ * \param size bytes to read
+ * \return value read
+ */
+static uint32_t read_be(FT_Byte *addr, size_t size)
+{
+    assert(size <= 4);
+
+    uint32_t ret = 0;
+    for (size_t i = 0; i < size; i++)
+        ret = (ret << 8) | addr[i];
+
+    return ret;
+}
+
+struct Axis {
+    uint16_t index;
+    uint32_t tag;
+    uint16_t ordering;
+    uint16_t fvar_idx;
+};
+
+struct AxisValue {
+    uint16_t axis_index;
+    int32_t value;
+    int32_t minValue;
+    int32_t maxValue;
+};
+
+struct NamedValueGroup {
+    uint16_t flags;
+    uint16_t name_id;
+    uint16_t nb_values;
+    struct AxisValue *values;
+};
+
+struct STAT {
+    size_t nb_axes;
+    size_t nb_value_groups;
+    struct Axis *axes;
+    struct Axis *sorted_axes;
+    struct NamedValueGroup *value_groups;
+    int elidedFallbackNameID;
+};
+
+static int compare_axes(const void *pa, const void *pb)
+{
+    const struct Axis *a = pa;
+    const struct Axis *b = pb;
+
+    return (int)a->ordering - (int)b->ordering;
+}
+
+#define STAT_FLAG_ELIDABLE_ALL 0x02
+#define STAT_FLAG_ELIDABLE_FAMILY 0x04 // libass-internal
+
+/**
+ * \brief Parse an OpenType STAT table.
+ *
+ * \param stat structure to fill
+ * \param data pointer to raw table data
+ * \param data_size size of data
+ * \param variants FreeType fvar data
+ * \return success
+ */
+static bool parse_stat(struct STAT *stat, FT_Byte *data, FT_ULong data_size, FT_MM_Var* variants)
+{
+    if (data_size < 20)
+        return false;
+
+    uint16_t majorVersion = read_be(data, 2);
+    uint16_t minorVersion = read_be(data + 2, 2);
+
+    uint16_t axis_size = read_be(data + 4, 2);
+    stat->nb_axes = read_be(data + 6, 2);
+    uint32_t axes_offset = read_be(data + 8, 4);
+
+    stat->nb_value_groups = read_be(data + 12, 2);
+    uint32_t values_offset = read_be(data + 14, 4);
+
+    if (majorVersion > 1 || minorVersion >= 1)
+        stat->elidedFallbackNameID = read_be(data + 18, 2);
+    else
+        stat->elidedFallbackNameID = -1;
+
+    if (axis_size < 8)
+        return false;
+
+    if (axes_offset + stat->nb_axes * axis_size > data_size)
+        return false;
+
+    if (values_offset + stat->nb_value_groups * 2 > data_size)
+        return false;
+
+    stat->axes = calloc(stat->nb_axes, sizeof(struct Axis));
+    stat->value_groups = calloc(stat->nb_value_groups, sizeof(struct NamedValueGroup));
+    if (!stat->axes)
+        return false;
+    if (!stat->value_groups)
+        return false;
+
+    int i, j;
+
+    for (i = 0; i < stat->nb_axes; i++) {
+        FT_Byte *p = data + axes_offset + i * axis_size;
+        struct Axis *a = &stat->axes[i];
+
+        a->index = i;
+        a->tag = read_be(p, 4);
+        a->ordering = read_be(p + 6, 2);
+
+        bool matched = false;
+
+        for (j = 0; j < variants->num_axis; j++) {
+            if (a->tag == variants->axis[j].tag) {
+                a->fvar_idx = j;
```

Comment:
This is incorrect.
For example, [Alegreya - Original.zip](https://github.com/user-attachments/files/15962083/Alegreya.-.Original.zip)
has 2 DesignAxisRecord (wght and ital) in the STAT table, but the fvar table only has 1 axis (wght).

Reply:
The sample font alone doesn't tell me much without knowing how it behaves in GDI. What's this supposed to do?